File: Data Structures & Algorithms/valid-tree/submission-4.java
class Solution {
    public boolean validTree(int n, int[][] edges) {
        if (edges.length > n - 1) {
            return false;
        }
        List<Integer>[] g = new ArrayList[n];
        Arrays.setAll(g, i -> new ArrayList<>());

        for (int [] e : edges) {
            g[e[0]].add(e[1]);
            g[e[1]].add(e[0]);
        }

        int[] visit = new int[n];
        
        if (!dfs(0, -1, visit, g)) {
            return false;
        }
        for (int v : visit) {
          if (v != 2) return false;
        }
        return true;
    }

    private boolean dfs(int i, int parent, int[] visit, List<Integer>[] g) {
        visit[i] = 1;
    
        for (int j : g[i]) {
            if (j == parent) {
                continue;
            }
            if (visit[j] == 1 || visit[j] == 0 && !dfs(j, i, visit, g)) {
                return false;
            }
        }
        visit[i] = 2;
        return true;
    }
}

File: Data Structures & Algorithms/count-connected-components/submission-4.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        g = [[] for _ in range(n)]
        for e1, e2 in edges:
            g[e1].append(e2)
            g[e2].append(e1)

        visit = [False] * n

        def dfs(i) -> int:
            visit[i] = True
            for v in g[i]:
                if not visit[v]:
                    dfs(v)
        
        ans = 0

        for i, b in enumerate(visit):
            if not b: 
                dfs(i)
                ans += 1
        return ans            

